dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java

```diff
@@ -360,29 +450,39 @@ void readBlock() throws HpccFileException
                 return;
             }
 
-            // Compact the array before reading in new data
-            int remainingBytesInBuffer = this.readBufferLen - this.readPos;
-            this.readBufferLen = remainingBytesInBuffer + dataLen;
+            while (dataLen > 0)
+            {
+                try
+                {
+                    this.bufferWriteMutex.acquire();
+                }
+                catch(Exception e){}
+
+                int totalBufferCapacity = this.readBuffer.length;
+                int currentBufferLen = this.readBufferLen.get();
+                int remainingBufferCapacity = totalBufferCapacity - currentBufferLen;
 
-            // Expand read buffer if necessary
-            if (this.readBufferLen > this.readBuffer.length) {
-                byte[] newBuffer = new byte[this.readBufferLen];
-                System.arraycopy(this.readBuffer,this.readPos,newBuffer,0,remainingBytesInBuffer);
-                this.readBuffer = newBuffer;
-            } else {
-                System.arraycopy(this.readBuffer,this.readPos,this.readBuffer,0,remainingBytesInBuffer);
-            }
+                remainingBufferCapacity = totalBufferCapacity - currentBufferLen;
+                int bytesToRead = Math.min(remainingBufferCapacity,Math.min(this.dis.available(),dataLen));
 
-            // Update the markPos
-            if (this.markPos >= 0) {
-                this.markPos -= this.readPos;
-            }
+                dis.readFully(this.readBuffer, currentBufferLen, bytesToRead);
 
-            if (dataLen > this.readBuffer.length) {
-                throw new HpccFileException("Error during read block, available data len is larger than read buffer. This should not happen");
+                // ReadBufferLen should only ever be updated within the mutex so no need to do compare exchange
+                this.readBufferLen.set(currentBufferLen + bytesToRead);
+                dataLen -= bytesToRead;
+
+                bufferWriteMutex.release();
+
+                // If we don't have room in the buffer to fetch more data sleep
+                if (dataLen > 0  && this.readBufferLen.get() > (BUFFER_PREFETCH_THRESHOLD_KB * 1024))
```

Comment:
do we need to ensure another thread hasn't written to the buffer, but hasn't set the readbufferlen yet (line 471)?